api-resources.md

```diff
@@ -0,0 +1,337 @@
+## Resources
+
+DAP is structured as an HTTP-based application with the following resources:
+
+- (TODO) Tasks
+- HPKE configurations
+- Reports
+- Aggregation jobs
+- Aggregate shares
+- Collections
+
+A resource's path is resolved relative to a server's base URL to construct a resource URI. Deployments may host resource paths arbitrarily deep relative to their domain. Paths are generally structured as `/resource-type/{resource-id}`. Anywhere `{resource-id}` (e.g., `{task-id}` or `{report-id}`) occurs in a URI is to be understood as the URL-safe, unpadded base64 representation of the resource's identifier, which itself is usually 16 random bytes (some are 32, but https://github.com/ietf-wg-ppm/draft-ietf-ppm-dap/issues/349 will normalize them to 16).
+
+Some resources are owned by another resource, and so their URI will contain two identifiers. For instance, a report belongs to a task, so the URI is `[base]/tasks/{task-id}/reports/{report-id}`.
+
+Generally, if a resource supports the `GET` method, it also supports `HEAD`. The response to `HEAD` is exactly the same as `GET`, but the response contains no body.
+
+If an HTTP method is listed in a resource's required HTTP methods, implementations MUST provide responses as indicated. Otherwise, implementations MAY implement methods on resources however they wish, including returning an error like HTTP 405 Method Not Allowed.
+
+We use HTTP `PUT` requests for idempotent creation of resources. Server responses to `PUT` requests follow [RFC 9110 section 9.3.4](https://httpwg.org/specs/rfc9110.html#rfc.section.9.3.4).
+
+Finally, a quick reminder of some HTTP semantics: `GET` and `PUT` are idempotent (i.e., the same request may be made multiple times without changing state in the server more than once, and the response will be the same each time). `POST` is not idempotent. `GET` requests may not have bodies.
+
+We care a lot about idempotence because we need to account for cases where a client makes a request but never sees the response, and thus needs a way to recover from this state of uncertainty.
+
+### Tasks
+
+TODO: do we want a task resource that we can GET and PUT in the protocol? Does this collide with https://wangshan.github.io/draft-wang-ppm-dap-taskprov/draft-wang-ppm-dap-taskprov.html or https://github.com/divergentdave/draft-dcook-ppm-dap-interop-test-design?
+
+### HPKE configurations
+
+#### Path
+
+`/hpke_config[?task_id=task-id]`
+
+The `task_id` query parameter is optional, as described in https://www.ietf.org/archive/id/draft-ietf-ppm-dap-02.html#section-4.3.1.
+
+##### Alternatives
+
+An HPKE config has an ID, so we could use a path `hpke_configs/hpke-config-id`. However, a client should never need to get any HPKE config except for the "current" one, so does it make sense to expand the API surface with a path param for HPKE config ID?
```

Comment:
As ChrisW would no doubt ask at this juncture: What are your requirements here? Why would we need multiple HPKE configs?

One reason was previously pointed out by @ekr: https://github.com/ietf-wg-ppm/draft-ietf-ppm-dap/issues/248. TL;DR: The Client and Aggregator may not initially agree on what HPKE ciphersuite to use. But we may not need to expose multiple configs simultaneously, if the Client can indicate in its request which suites it supports.

Reply:
This change will be complicated enough, so I think I will choose not to solve #248, too, so I lean toward leaving the `hpke_config` exactly as it is.

- [ ] put `hpke_config` back the way it was